File: ios/Arculus/CSDK/WalletInitSession.swift
import CSDK

final class WalletInitSession: CSDKAPICall<Void> {
    override func request() async throws -> [Data] {
        var len: size_t = 0
        
        guard let bytes = WalletInitSessionRequest(wallet, &len) else {
            throw CSDKAPICallError(.requestCreationFailed("init session"))
        }
        
        return [Data(bytes: bytes, count: len)]
    }
    
    override func response(bytes: [UInt8]) async throws -> ResponseType {
        let result = WalletInitSessionResponse(wallet, bytes, bytes.count)
        
        try CSDKAPICallError.validateWalletResponseCode(rc: result)
    }
}

File: android/src/main/java/com/aptoslabs/arculus/csdk/WalletInitSession.kt
package com.aptoslabs.arculus.csdk

import com.sun.jna.Pointer

class WalletInitSession(wallet: Pointer) : CSDKAPICall<Unit>(wallet) {
  override suspend fun request(): Array<ByteArray?> {
    val len = SizeTByReference()

    val pointer = CSDK.WalletInitSessionRequest(wallet, len)

    if (pointer == Pointer.NULL) throw CSDKAPICallRequestCreationException()

    return arrayOf(pointer.getByteArray(0, len.toIntChecked()))
  }

  override suspend fun response(bytes: ByteArray) {
    val rc = CSDK.WalletInitSessionResponse(wallet, bytes, bytes.size)

    validateWalletResponse(rc)
  }
}
